cmake/helpers.cmake

```diff
@@ -138,18 +136,34 @@ macro(add_sanitizer_flag flag)
     # Check C and CXX compilers for given sanitizer flag.
     check_c_compiler_flag("${SANITIZER_FLAG}" "C_${check_name}")
     check_cxx_compiler_flag("${SANITIZER_FLAG}" "CXX_${check_name}")
-    if (${C_${check_name}} OR ${CXX_${check_name}})
-        # Set appropriate linker flags for building executables.
-        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${SANITIZER_FLAG} ${SANITIZER_ARGS}")
-        if (${C_${check_name}})
-            set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${SANITIZER_FLAG} ${SANITIZER_ARGS}")
-        endif()
-        if (${CXX_${check_name}})
-            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${SANITIZER_FLAG} ${SANITIZER_ARGS}")
-        endif()
+    if (NOT ${C_${check_name}} OR NOT ${CXX_${check_name}})
+        message(FATAL_ERROR "${flag} sanitizer is not supported (either by C or CXX compiler)")
+    endif()
+
+    # Check C and CXX compilers for sanitizer arguments.
+    if (${SANITIZER_ARGS})
+        check_c_compiler_flag("${SANITIZER_ARGS}" "C_HAS_SAN_ARGS")
+        check_cxx_compiler_flag("${SANITIZER_ARGS}" "CXX_HAS_SAN_ARGS")
+    endif()
+
+    if (${C_HAS_SAN_ARGS} AND ${CXX_HAS_SAN_ARGS})
+        set(SANITIZER_OPTION "${SANITIZER_FLAG} ${SANITIZER_ARGS}")
+    elseif (${SANITIZER_ARGS})
+        # Ignore sanitizer argument as it's not supported but was set.
+        message(WARNING "sanitizer argument ${SANITIZER_ARGS} is not supported (either by C or CXX compiler)")
+        set(SANITIZER_OPTION "${SANITIZER_FLAG}")
     else()
-        message(FATAL_ERROR "${flag} sanitizer is not supported (neither by C nor CXX compiler)")
+        # No sanitizer argument was set.
+        set(SANITIZER_OPTION "${SANITIZER_FLAG}")
     endif()
 
-    set(CMAKE_REQUIRED_LIBRARIES ${SAVED_CMAKE_REQUIRED_LIBRARIES})
+    add_compile_options("${SANITIZER_OPTION}")
+
+    # Clang/gcc needs the flag added to the linker. The Microsoft LINK linker doesn't recognize
+    # sanitizer flags and will give a LNK4044 warning.
+    if (NOT MSVC)
+        add_link_options("${SANITIZER_OPTION}")
+	endif()
```

Comment:
pls align indentation for `endif` and `if`

Reply:
Done.